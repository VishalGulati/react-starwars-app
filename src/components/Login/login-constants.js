export const LOGIN_ACTION = 'LOGIN';
export const LOGIN_SUCCESS_ACTION = 'LOGIN_SUCCESS';
export const LOGIN_FAIL_ACTION = 'LOGIN_FAIL';
