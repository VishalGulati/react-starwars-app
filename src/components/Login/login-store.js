import { put, takeLatest, call } from 'redux-saga/effects';
import {
  LOGIN_ACTION,
  LOGIN_SUCCESS_ACTION,
  LOGIN_FAIL_ACTION
} from './login-constants';
import { authenticate } from '../../api';

const initialState = {
  isLoading: false,
  data: null,
  error: null
};

export const loginAction = ({ username, password }) => ({
  type: LOGIN_ACTION,
  username,
  password
});

export const loginSuccess = data => ({
  type: LOGIN_SUCCESS_ACTION,
  data
});

export const loginFail = error => ({
  type: LOGIN_FAIL_ACTION,
  error
});

export function* loginSaga(action) {
  try {
    const { username, password } = action;
    const data = yield call(authenticate, { username, password });
    yield put(loginSuccess(data));
  } catch ({ code, message }) {
    yield put(loginFail({ code, message }));
  }
}

export function* watchLogin() {
  yield takeLatest(LOGIN_ACTION, loginSaga);
}

export default function loginReducer(state = initialState, action) {
  switch (action.type) {
    case LOGIN_ACTION:
      return {
        ...state,
        isLoading: true
      };

    case LOGIN_SUCCESS_ACTION:
      return {
        ...state,
        isLoading: false,
        data: action.data
      };

    case LOGIN_FAIL_ACTION:
      return {
        ...state,
        isLoading: false,
        error: action.error
      };

    default:
      return state;
  }
}
