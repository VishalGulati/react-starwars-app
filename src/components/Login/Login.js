import React, { Component } from 'react';
import {
    Col,
    Button
} from 'react-bootstrap';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { loginAction } from './login-store';
import { searchRoute } from '../../router/route-constants';

export class Login extends Component {
    constructor(props) {
        super(props);
        this.state = {
            username: '',
            password: ''
        };
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.login !== this.props.login) {
            const { isLoading, error } = nextProps.login;
            if (!isLoading && !error) {
                this.props.history.push(`/${searchRoute}`);
            }
        }
    }

    getValidationState = () => {
        const { username, password } = this.state;
        if (username && password) return 'success';
        else return 'error';
    };

    handleUsernameChange = e => {
        this.setState({ username: e.target.value });
    };

    handlePasswordChange = e => {
        this.setState({ password: e.target.value });
    };

    userLogin = e => {
        const { username, password } = this.state;
        this.props.loginAction({ username, password });
        e.preventDefault();
    };

    render() {
        return (<div className="login container">
            <h1 className="text-center" style={{ marginTop: '10px' }}> Login </h1>
            <form className="form-horizontal">
                <div className="form-group">
                    <label className="control-label col-sm-2" htmlFor="email">Email:</label>
                    <div className="col-sm-10">
                        <input type="email" className="form-control" id="email" placeholder="Enter email" name="email"
                            onChange={this.handleUsernameChange}
                        />
                    </div>
                </div>

                <div className="form-group">
                    <label className="control-label col-sm-2" htmlFor="pwd">Password:</label>
                    <Col sm={10} >
                        <input type="password" className="form-control" id="pwd" placeholder="Enter password" name="pwd"
                            onChange={this.handlePasswordChange}
                        />
                    </Col>
                </div>

                <div className="form-group">
                    <Col smOffset={2}
                        sm={10} >
                        <Button bsStyle="primary"
                            type="submit"
                            onClick={this.userLogin} >
                            Login
                        </Button>
                    </Col>
                </div>
            </form >
        </div >
        );
    }
}

const mapStateToProps = ({ login }) => ({
    login
});

const mapDispatchToProps = dispatch =>
    bindActionCreators({ loginAction },
        dispatch
    );

export default connect(mapStateToProps, mapDispatchToProps)(Login);
