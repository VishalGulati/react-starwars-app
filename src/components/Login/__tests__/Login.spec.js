import React from 'react';
import renderer from 'react-test-renderer';
import { Login } from '../Login';

describe('Login', () => {
    it('should render Login properly', () => {
        const component = renderer.create(<Login />);
        expect(component.toJSON()).toMatchSnapshot();
    });
});
