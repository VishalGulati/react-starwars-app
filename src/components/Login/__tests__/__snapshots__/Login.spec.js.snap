// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render Login properly 1`] = `
<div
  className="login"
>
  <div
    className="text-center page-header"
  >
    <h1>
       Login 
    </h1>
  </div>
  <form
    className="form-horizontal"
  >
    <div
      className="form-group"
    >
      <label
        className="col-sm-2 control-label"
        htmlFor="formUsername"
      >
        Email
      </label>
      <div
        className="col-sm-10"
      >
        <input
          className="form-control"
          id="formUsername"
          onChange={[Function]}
          type="text"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <label
        className="col-sm-2 control-label"
        htmlFor="form  Password"
      >
        Password
      </label>
      <div
        className="col-sm-10"
      >
        <input
          className="form-control"
          id="form  Password"
          onChange={[Function]}
          type="password"
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <div
        className="col-sm-10 col-sm-offset-2"
      >
        <button
          className="btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="submit"
        >
          Login
        </button>
      </div>
    </div>
  </form>
</div>
`;
